Libs/MRML/Core/vtkMRMLScene.cxx

```diff
@@ -1285,14 +1284,61 @@ vtkMRMLNode* vtkMRMLScene::AddNewNodeByClass(
   return this->AddNode(nodeToAdd);
 }
 
+//------------------------------------------------------------------------------
+vtkMRMLNode* vtkMRMLScene::AddNewNodeByClassWithID(std::string className, std::string nodeBaseName, std::string nodeID)
+{
+  if (className.empty())
+    {
+    vtkErrorMacro("AddNewNodeByClassWithID: className is an emptry string");
+    return NULL;
+    }
+
+  if (nodeID.empty())
+    {
+    vtkErrorMacro("AddNewNodeByClassWithID: nodeID is an emptry string");
+    return NULL;
+    }
+
+  bool isUnique = ((this->GetNodeByID(nodeID) == 0) &&
+                   (!this->IsReservedID(nodeID)) &&
+                   (!this->IsNodeIDReservedByUndo(nodeID)));
+  if (!isUnique)
+    {
+    vtkErrorMacro("AddNewNodeByClassWithID: nodeID is already in use");
+    return NULL;
+    }
+
+  if (this->NodeIDs.find(nodeID) != this->NodeIDs.end())
+    {
+    vtkErrorMacro("AddNewNodeByClassWithID: node already exists with ID - " << nodeID);
+    return NULL;
+    }
+
+  vtkSmartPointer<vtkMRMLNode> nodeToAdd =
+    vtkSmartPointer<vtkMRMLNode>::Take(this->CreateNodeByClass(className.c_str()));
+  if (nodeToAdd == NULL)
+    {
+    vtkErrorMacro("AddNewNodeByClassWithID: failed to create node by class " << className);
+    return NULL;
+    }
+
+  nodeToAdd->SetID(nodeID.c_str());
+  if (!nodeBaseName.empty())
+    {
+    nodeToAdd->SetName(nodeBaseName.c_str());
+    }
+
+  return this->AddNode(nodeToAdd);
```

Comment:
I wonder if we should explicitly update the `UniqueIDs` map so that `GetUniqueIDIndex` doesn't accidentally override an ID explicitly provided.

Reply:
I just realized this is already taken care of in the modified `vtkMRMLScene::GetUniqueIDIndex`.